spock-core/src/main/java/spock/util/concurrent/PollingConditions.java

```diff
@@ -148,34 +148,43 @@ public void eventually(Closure<?> conditions) throws InterruptedException {
    * @throws InterruptedException if evaluation is interrupted
    */
   @ConditionBlock
-  public void within(double seconds, Closure<?> conditions) throws InterruptedException  {
+  public void within(double seconds, Closure<?> conditions) throws InterruptedException {
     long timeoutMillis = toMillis(seconds);
     long start = System.currentTimeMillis();
     long lastAttempt = 0;
-    Thread.sleep(toMillis(initialDelay));
+    // Thread.sleep(0) would yield the processor to higher-priority threads,
+    // which would slow down a polling condition without initial delay unnecessarily
+    if (initialDelay > 0) {
+      Thread.sleep(toMillis(initialDelay));
+    }
 
     long currDelay = toMillis(delay);
     int attempts = 0;
+    long elapsedTime = 0;
+    Throwable testException = null;
 
-    while(true) {
+    while (elapsedTime <= timeoutMillis) {
       try {
-        attempts++;
-        lastAttempt = System.currentTimeMillis();
-        GroovyRuntimeUtil.invokeClosure(conditions);
+        try {
+          GroovyRuntimeUtil.invokeClosure(conditions);
+        } finally {
+          lastAttempt = System.currentTimeMillis();
```

Comment:
Actually, you are right, `lastAttempt` (which should really have been named `previousAttempt`, BTW) can go away because the `elapsedTime` calculation needs to go to the `finally` block. The time needs to be updated after evaluating the condition, no matter if there was an exception while evaluting the condition or not.

I prefer to increase the `attempts` counter after the condition was evaluated, not before, though. It is a matter of taste because we know that even when increasing it before the evaluation, same evaluation will happen next anyway. But to me it feels more "transactional" to do it afterwards.

Reply:
Just saw your comment under the test

> `eventually` only retries when an exception occurs

Yes, of course! I am so stupid! There is an unconditional `return` after the condition is evaluated without exception and the conditions or asserts inside the closure implicitly throw "condition not satisfied" or assertion errors. They don't just return `true` or `false`. I was thinking in the wrong direction, OMG. So my changes complicated the code more than necessary. I will simplify it to just avoid the `Thread.sleep(0)` and the `while (true)`. Thanks for your code review and for challenging me.